src/utils/cache.rs

```diff
@@ -62,20 +62,65 @@ pub trait CacheExt: Cache {
         timeout: usize,
-    ) -> Result<String> {
-        let data: String =
-            match self.fetch(&url) {
-                Some(cached) => cached,
-                None => {
-                    let response = client.get(url).send()?;
-                    // Don't cache if it is a Server error
-                    if response.status().is_server_error() { anyhow::bail!("Got server error for {}", url); };
-                    let raw_data = response.text()?;
-                    self.create(&url, &raw_data, timeout);
-                    raw_data
+    ) -> ApiResult<String> {
+        match self.fetch(&url) {
+            Some(cached) => {
+                let cached_with_code = CachedWithCode::split(&cached);
+                if !cached_with_code.is_error() {
+                    Ok(String::from(cached_with_code.data))
+                } else {
+                    return if let Ok(backend_error) = serde_json::from_str::<BackendError>(&cached_with_code.data) {
+                        Err(ApiError { status: cached_with_code.code, message: ApiErrorMessage::BackendError(backend_error) })
+                    } else {
+                        Err(ApiError { status: cached_with_code.code, message: ApiErrorMessage::SingleLine(cached_with_code.data) })
+                    };
                 }
-            };
-        Ok(data)
+            }
+            None => {
+                let response = client.get(url).send()?;
+                if response.status().is_server_error() {
+                    return Err(anyhow::anyhow!("Got server error for {}", url).into());
+                }
+                let status_code = response.status().as_u16();
+                let is_client_error = response.status().is_client_error();
+                let raw_data = response.text()?;
+                self.create(&url, CachedWithCode::join(status_code, &raw_data).as_str(), timeout);
+                if is_client_error {
+                    return if let Ok(backend_error) = serde_json::from_str::<BackendError>(&raw_data) {
+                        Err(ApiError { status: status_code, message: ApiErrorMessage::BackendError(backend_error) })
+                    } else {
+                        Err(ApiError { status: status_code, message: ApiErrorMessage::SingleLine(raw_data) })
+                    };
+                }
+                Ok(raw_data)
+            }
+        }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub(super) struct CachedWithCode {
+    pub(super) code: u16,
+    pub(super) data: String,
+}
+
+impl CachedWithCode {
+    const SEPARATOR: &'static str = ";";
+
+    pub(super) fn split(cached: &str) -> Self {
+        let cached_with_code: Vec<&str> = cached.split(CachedWithCode::SEPARATOR).collect();
+        CachedWithCode {
+            code: cached_with_code.get(0).expect("Must have a status code").parse().expect("Not a valid Http code"),
+            data: cached_with_code.get(1).expect("Must have data").to_string(),
```

Comment:
We should take the rest not just the second part right? E.g. if we cache the following `404:{"this_is_some_json":"And here we have a message with ;"}` then then we would have 3 parts and your `data` part would be invalid.